src/spclient.rs

```diff
@@ -163,6 +201,365 @@ impl SpClient {
+                prefix: PSBT_SP_PREFIX.as_bytes().to_vec(),
+                subtype: PSBT_SP_SUBTYPE,
+                key: PSBT_SP_TWEAK_KEY.as_bytes().to_vec()
+            }, tweak.to_be_bytes().to_vec());
+            psbt.inputs[i] = psbt_input;
+        }
+        
+        for (i, recipient) in recipients.iter().enumerate() {
+            if let Ok(sp_address) = SilentPaymentAddress::try_from(recipient.address.as_str()) {
+                // Add silentpayment address to the output
+                let mut psbt_output = Output { ..Default::default() };
+                psbt_output.proprietary.insert(raw::ProprietaryKey {
+                    prefix: PSBT_SP_PREFIX.as_bytes().to_vec(),
+                    subtype: PSBT_SP_SUBTYPE,
+                    key: PSBT_SP_ADDRESS_KEY.as_bytes().to_vec()
+                }, serialize(&sp_address.to_string()));
+                psbt.outputs[i] = psbt_output;
+            } else {
+                // Regular address, we don't need to add more data
+                continue;
+            }
+        }
+
+        Ok(psbt)
+    }
+
+    pub fn get_a_sum_secret_keys(input: &Vec<SecretKey>) -> SecretKey {
+        let secp = Secp256k1::new();
+
+        let mut negated_keys: Vec<SecretKey> = vec![];
+
+        for key in input {
+            let (_, parity) = key.x_only_public_key(&secp);
+
+            if parity == bitcoin::secp256k1::Parity::Odd {
+                negated_keys.push(key.negate());
+            } else {
+                negated_keys.push(key.clone());
+            }
+        }
+
+        let (head, tail) = negated_keys.split_first().unwrap();
+
+        let result: SecretKey = tail
+            .iter()
+            .fold(*head, |acc, &item| acc.add_tweak(&item.into()).unwrap());
+
+        result
+    }
+
+    fn taproot_sighash<T: std::ops::Deref<Target = Transaction> + std::borrow::Borrow<Transaction>> (
+            input: &Input,
+            prevouts: &Vec<&TxOut>,
+            input_index: usize,
+            cache: &mut SighashCache<T>,
+            tapleaf_hash: Option<TapLeafHash>
+        ) -> Result<(Message, PsbtSighashType), Error> {
+        let prevouts = Prevouts::All(prevouts);
+
+        let hash_ty = input
+            .sighash_type
+            .map(|ty| ty.taproot_hash_ty())
+            .unwrap_or(Ok(bitcoin::TapSighashType::Default))?;
+
+        let sighash = match tapleaf_hash {
+            Some(leaf_hash) => cache.taproot_script_spend_signature_hash(
+                input_index,
+                &prevouts,
+                leaf_hash,
+                hash_ty,
+            )?,
+            None => cache.taproot_key_spend_signature_hash(input_index, &prevouts, hash_ty)?,
+        };
+        let msg = Message::from_digest(sighash.into_32());
+        Ok((msg, hash_ty.into()))
+    }
+
+    // Sign a transaction with garbage, used for easier fee estimation
+    fn sign_psbt_fake(psbt: &Psbt) -> Transaction {
```

Comment:
maybe we can skip making a fake psbt later on, by guessing the tx weight on our inputs/outputs